### PR TITLE
Permit 'R'/'r' format specifier with integer formatting

### LINKED ISSUE #67623: Making numerical formatting specifiers more consistent
Today 'R'/'r' aren't supported with integral types but are with floating point, and 'X'/'x' aren't supported with floating point but are with integral types.  Related to both https://github.com/dotnet/runtime/issues/65634 and the generic math support, where `INumber<T>` inherits `ISpanFormattable` and thus we should expect format specifiers to be used with arbitrary numerical types, we should allow all of these specifiers to be used, regardless of integer vs floating point.  'R'/'r' for integers can just be treated as 'G'/'g', respectively, and 'X'/'x' for floating point is covered by https://github.com/dotnet/runtime/issues/1630.

(Discussed in API review on 4/5/2022.)

cc: @tannergooding

### PR DESCRIPTION
**Pull request by stephentoub**
Treat it just like 'G'/'g'.  This is to minimize differences in numerical format allowances, primarily in support of generic math.

Fixes https://github.com/dotnet/runtime/issues/67623

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Utf8Formatter/Utf8Formatter.Integer.Unsigned.cs [hunk 1]
```diff
@@ -25,8 +25,12 @@ private static bool TryFormatUInt64(ulong value, Span<byte> destination, out int
             {
                 case 'G':
                 case 'g':
+                case 'R':
+                case 'r': // treat 'R'/'r' (which historically was only for floating-point) as being equivalent to 'g', rather than throwing
                     if (format.HasPrecision)
+                    {
                         throw new NotSupportedException(SR.Argument_GWithPrecisionNotSupported); // With a precision, 'G' can produce exponential format, even for integers.
```
Comment (tannergooding) [inline src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Utf8Formatter/Utf8Formatter.Integer.Unsigned.cs:32]: Hmmm. Preventing this seems wrong and inconsistent with what `UTF16` formatting does. (`5.ToString("G99")` is valid)

The official docs (https://docs.microsoft.com/en-us/dotnet/standard/base-types/standard-numeric-format-strings) cover that it produces `Result: The more compact of either fixed-point or scientific notation` and `Precision specifier: Number of significant digits.`

There are notably some other inconsistencies here as well. Integers are missing support for `C`, `E`, `F`, and `P`. `Floating-point` more broadly doesn't support `D` either, but reasonably could (its basically just an alternative to `F0`)

Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Utf8Formatter/Utf8Formatter.Integer.Unsigned.cs:32]: -- Noting this is existing and I don't think this PR needs to address it, just calling it out as something else that might need broader consideration.

I didn't realize the UTF8 formatter was "so limited" compared to the UTF16 one.

**Review by tannergooding — APPROVED** [review]

File: src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Utf8Formatter/Utf8Formatter.Integer.Unsigned.cs (hunk 1 above)
Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Utf8Formatter/Utf8Formatter.Integer.Unsigned.cs:32]: Thanks.  Mind opening a separate issue to track that then?  It sounds like it's something we should address (at some point).

Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Utf8Formatter/Utf8Formatter.Integer.Unsigned.cs:32]: https://github.com/dotnet/runtime/issues/67962

**Comment by danmoseley** [comment]
https://github.com/dotnet/runtime/issues/67878

**Comment by dakersnar** [comment]
Might be the cause of windows x86 performance improvements: https://github.com/dotnet/perf-autofiling-issues/issues/4710

**Comment by stephentoub** [comment]
> Might be the cause of windows x86 performance improvements: https://github.com/dotnet/perf-autofiling-issues/issues/4710

Much more likely it's https://github.com/dotnet/runtime/pull/67902.

**Pull request merged**